### PR TITLE
matrix-appservice-irc: init at 0.25.0

### PR DESCRIPTION
**Pull request by piegamesde**
###### Motivation for this change

It's time for iteration 9 (:

This PR:

- Starts with @JJJollyjim's #95854
- Does not update to the latest version. I don't want to introduce any more potential merge blockers. An update can easily be done in a future PR.
- Fixes a few issues from the review of the previous (see the respective commit message).

Thanks to all those whose work this is building on.

CC @mweinelt @tklitschi

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/matrix-synapse/matrix-appservice-irc/node-composition.nix [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# This file has been generated by node2nix 1.8.0. Do not edit!
```
Comment (SuperSandro2000) [inline pkgs/servers/matrix-synapse/matrix-appservice-irc/node-composition.nix:1]: Please use node2nix 1.9.0 which has some important changes required to fix eval.

File: pkgs/servers/matrix-synapse/matrix-appservice-irc/node-composition.nix [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  nodeEnv = import ../../../development/node-packages/node-env.nix {
+    inherit (pkgs) stdenv python2 utillinux runCommand writeTextFile;
+    inherit nodejs;
+    libtool = if pkgs.stdenv.isDarwin then pkgs.darwin.cctools else null;
+  };
+in
+import ./node-packages.nix {
+  inherit (pkgs) fetchurl fetchgit;
+  inherit nodeEnv;
+}
```
Comment (SuperSandro2000) [inline pkgs/servers/matrix-synapse/matrix-appservice-irc/node-composition.nix:17]: Missing final newline.

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -4910,6 +4910,8 @@ in
 
   matrix-synapse-plugins = recurseIntoAttrs matrix-synapse.plugins;
 
+  matrix-appservice-irc = callPackage ../servers/matrix-synapse/matrix-appservice-irc {};
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:4913]:
```suggestion
  matrix-appservice-irc = callPackage ../servers/matrix-synapse/matrix-appservice-irc { };
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/servers/matrix-synapse/matrix-appservice-irc/default.nix [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    inherit (stdenv.hostPlatform) system;
+  };
+in
+ourNodePackages."${packageName}".override {
+  nativeBuildInputs = [ makeWrapper ];
+  buildInputs = [ nodePackages.node-gyp-build ];
+
+  postInstall = ''
+    makeWrapper '${nodejs}/bin/node' "$out/bin/matrix-appservice-irc" \
+    --add-flags "$out/lib/node_modules/matrix-appservice-irc/app.js"
```
Comment (SuperSandro2000) [inline pkgs/servers/matrix-synapse/matrix-appservice-irc/default.nix:18]:
```suggestion
      --add-flags "$out/lib/node_modules/matrix-appservice-irc/app.js"
```

File: pkgs/servers/matrix-synapse/matrix-appservice-irc/default.nix [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  '';
+
+  meta = with lib; {
+    description = "Node.js IRC bridge for Matrix";
+    maintainers = with maintainers; [ puffnfresh jjjollyjim ];
+    homepage = "https://github.com/matrix-org/matrix-appservice-irc";
+    license = licenses.asl20;
+  };
+
+  passthru.tests.matrix-appservice-irc = nixosTests.matrix-appservice-irc;
```
Comment (SuperSandro2000) [inline pkgs/servers/matrix-synapse/matrix-appservice-irc/default.nix:28]:
```suggestion
```

File: pkgs/servers/matrix-synapse/matrix-appservice-irc/default.nix [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+in
+ourNodePackages."${packageName}".override {
+  nativeBuildInputs = [ makeWrapper ];
+  buildInputs = [ nodePackages.node-gyp-build ];
+
+  postInstall = ''
+    makeWrapper '${nodejs}/bin/node' "$out/bin/matrix-appservice-irc" \
+    --add-flags "$out/lib/node_modules/matrix-appservice-irc/app.js"
+  '';
+
```
Comment (SuperSandro2000) [inline pkgs/servers/matrix-synapse/matrix-appservice-irc/default.nix:20]:
```suggestion

  passthru.tests.matrix-appservice-irc = nixosTests.matrix-appservice-irc;
 
```

File: pkgs/servers/matrix-synapse/matrix-appservice-irc/default.nix [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  packageName = with lib; concatStrings (map (entry: (concatStrings (mapAttrsToList (key: value: "${key}-${value}") entry))) (importJSON ./package.json));
+
+  ourNodePackages = import ./node-composition.nix {
+    inherit pkgs nodejs;
+    inherit (stdenv.hostPlatform) system;
+  };
+in
+ourNodePackages."${packageName}".override {
+  nativeBuildInputs = [ makeWrapper ];
+  buildInputs = [ nodePackages.node-gyp-build ];
```
Comment (SuperSandro2000) [inline pkgs/servers/matrix-synapse/matrix-appservice-irc/default.nix:14]: I think this is executed at build time so it should go to native.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/servers/matrix-synapse/matrix-appservice-irc/default.nix (hunk 7 above)
Comment (piegamesde) [reply pkgs/servers/matrix-synapse/matrix-appservice-irc/default.nix:14]: Somebody with NodeJS packaging experience please confirm this.

File: pkgs/servers/matrix-synapse/matrix-appservice-irc/default.nix (hunk 6 above)
Comment (piegamesde) [reply pkgs/servers/matrix-synapse/matrix-appservice-irc/default.nix:20]: What does this do?

File: pkgs/servers/matrix-synapse/matrix-appservice-irc/default.nix (hunk 7 above)
Comment (mweinelt) [reply pkgs/servers/matrix-synapse/matrix-appservice-irc/default.nix]: Just do it, run the test, if it succeeds and we are likely fine. :)

**Comment by Munksgaard** [comment]
Crossing my fingers for this one to come through!

**Comment by Munksgaard** [comment]
Is that an error in the passthru tests?

**Comment by SuperSandro2000** [comment]
I think so

@ofborg test matrix-appservice-irc

**Comment by Munksgaard** [comment]
I'm not even sure what those error messages mean...

**Comment by mweinelt** [comment]
> jsonschema.exceptions.SchemaError: [{'type': 'string', 'enum': ['admin']}] is not valid under any of the given schemas

This looks like an error in their schema validation, not sure. https://github.com/matrix-org/matrix-appservice-irc/blob/9f119e27cf9070ff1a5d9b22da3c0f9d4ca35a36/config.schema.yml#L137-L141

```patch
            permissions:
                type: "object"
                additionalProperties:
-                   - type: "string"
+                     type: "string"
                      enum: ["admin"]
```

**Comment by Munksgaard** [comment]
Of course, now you'll probably need to update to the latest version, if https://github.com/matrix-org/matrix-appservice-irc/pull/1261 is the fix...

**Comment by mweinelt** [comment]
> Of course, now you'll probably need to update to the latest version, if [matrix-org/matrix-appservice-irc#1261](https://github.com/matrix-org/matrix-appservice-irc/pull/1261) is the fix...

Just fetchpatch it for now until a new release bundles it.

**Comment by piegamesde** [comment]
As a new release has been announced for next week, I will skip the fetchpatch stage

**Comment by expipiplus1** [comment]
It would be really nice to see some content in the manual explaining this, perhaps with some kind of minimal example. (Wouldn't want this to gate the PR though!!!)

**Comment by piegamesde** [comment]
@puffnfresh @JJJollyjim Is anybody of you *not* comfortable with maintaining this package?

File: nixos/modules/services/misc/matrix-appservice-irc.nix [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+      default = 8009;
+    };
+
+    needBindingCap = mkOption {
+      type = bool;
+      description = "Whether the daemon needs to bind to ports below 1024 (e.g. for the ident service)";
+      default = false;
+    };
+
+    passwordKeyLength = mkOption {
```
Comment (mweinelt) [inline nixos/modules/services/misc/matrix-appservice-irc.nix:45]: passwordEncryptionKeyLength

File: nixos/modules/services/misc/matrix-appservice-irc.nix [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+
+    registrationUrl = mkOption {
+      type = str;
+      description = "The URL where the application service is listening for HS requests, from the Matrix HS perspective.";
+      example = "http://localhost:8009";
+    };
+
+    localpart = mkOption {
+      type = str;
+      description = "The user_id localpart to assign to the AS";
```
Comment (mweinelt) [inline nixos/modules/services/misc/matrix-appservice-irc.nix:60]: AS -> Application Service

File: nixos/modules/services/misc/matrix-appservice-irc.nix [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+    passwordKeyLength = mkOption {
+      type = ints.unsigned;
+      description = "Length of the key to encrypt IRC passwords";
+      default = 4096;
+      example = 8192;
+    };
+
+    registrationUrl = mkOption {
+      type = str;
+      description = "The URL where the application service is listening for HS requests, from the Matrix HS perspective.";
```
Comment (mweinelt) [inline nixos/modules/services/misc/matrix-appservice-irc.nix:54]: HS -> homeserver

File: nixos/modules/services/misc/matrix-appservice-irc.nix [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+    };
+
+    localpart = mkOption {
+      type = str;
+      description = "The user_id localpart to assign to the AS";
+      default = "appservice-irc";
+    };
+
+    settings = mkOption {
+      description = ''Configuration for the appservice, see <link xlink:href="https://github.com/matrix-org/matrix-appservice-irc/blob/${pkgs.matrix-appservice-irc.version}/config.sample.yaml"/> for supported values'';
```
Comment (mweinelt) [inline nixos/modules/services/misc/matrix-appservice-irc.nix:65]: Split this into multiple lines.

File: nixos/modules/services/misc/matrix-appservice-irc.nix [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+
+              options = {
+                url = mkOption {
+                  type = str;
+                  description = "The URL to the home server for client-server API calls";
+                };
+
+                domain = mkOption {
+                  type = str;
+                  description = ''The 'domain' part for user IDs on this home server. Usually (but not always) is the "domain name" part of the HS URL.'';
```
Comment (mweinelt) [inline nixos/modules/services/misc/matrix-appservice-irc.nix:85]: Split this into multiple lines.

File: nixos/modules/services/misc/matrix-appservice-irc.nix [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+                engine = mkOption {
+                  type = str;
+                  description = "Which database engine to use";
+                  default = "nedb";
+                };
+
+                connectionString = mkOption {
+                  type = str;
+                  description = "The database connection string";
+                  default = "nedb://var/lib/matrix-appservice-irc/data";
```
Comment (mweinelt) [inline nixos/modules/services/misc/matrix-appservice-irc.nix:107]: Can we get a postgresql example here?

File: nixos/modules/services/misc/matrix-appservice-irc.nix [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+            default = {};
+            description = "Configuration for the database";
+            type = submodule {
+              freeformType = jsonType;
+
+              options = {
+                engine = mkOption {
+                  type = str;
+                  description = "Which database engine to use";
+                  default = "nedb";
```
Comment (mweinelt) [inline nixos/modules/services/misc/matrix-appservice-irc.nix:101]: Can we add postgresql as example?

File: nixos/modules/services/misc/matrix-appservice-irc.nix [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+          ircService = mkOption {
+            default = {};
+            description = "IRC bridge configuration";
+            type = submodule {
+              freeformType = jsonType;
+
+              options = {
+                passwordEncryptionKeyPath = mkOption {
+                  type = str;
+                  description = "Location of the key with which IRC passwords are encrypted for storage. Will be generated on first run if not present.";
```
Comment (mweinelt) [inline nixos/modules/services/misc/matrix-appservice-irc.nix:122]: Split this into multiple lines.

File: nixos/modules/services/misc/matrix-appservice-irc.nix [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+    systemd.services.matrix-appservice-irc = {
+      description = "Matrix-IRC bridge";
+      before = [ "matrix-synapse.service" ]; # So the registration can be used by Synapse
+      wantedBy = [ "multi-user.target" ];
+
+      preStart = ''
+        umask 077
+        # Generate key for crypting passwords
+        if ! [ -f "${cfg.settings.ircService.passwordEncryptionKeyPath}" ]; then
+          ${pkgs.openssl}/bin/openssl genpkey -out "${cfg.settings.ircService.passwordEncryptionKeyPath}" -outform PEM -algorithm RSA -pkeyopt "rsa_keygen_bits:${toString cfg.passwordKeyLength}"
```
Comment (mweinelt) [inline nixos/modules/services/misc/matrix-appservice-irc.nix:147]: Split this into multiple lines.

**Comment by mweinelt** [comment]
Dear authors of previous iterators,

this journey is about to end. Thanks go out to everybody who made a contribution along the way.

This is it, for now. May there be many incremental improvements from here on!

@JJJollyjim @puffnfresh @gdamjan @pacien @dasJ @etu @Ralith

**Comment by expipiplus1** [comment]
The Wiki should be updated now that this is no longer "a WIP"! https://nixos.wiki/wiki/Matrix#matrix-appservice-irc

**Comment by piegamesde** [comment]
@expipiplus1 I'll take care of it soon™

**Comment by dasJ** [comment]
Science said it couldn't be done! It was said to be impossible to have matrix-appservice-irc in nixpkgs. But after all this time, @mweinelt proved that science was wrong here

**Pull request merged**